File: packages/nodejs/src/domains/auth/signout.ts
import { repositoryContainer } from '../../repositories';
import { getDebug } from '../../logging';

const debug = getDebug('domains:auth:jwt');

export interface RevokedToken {
  id: string;
  token: string;
  revokedAt: Date;
  createdAt: Date;
  updatedAt: Date;
}

export interface RevokedTokenCreateAttributes {
  token: string;
  revokedAt: Date;
}

export interface RevokedTokenUpdateAttributes {
  revokedAt: Date;
}

// サインアウト - tokenを無効化
export const signout = async (token?: string | null): Promise<boolean> => {
  if (!token) {
    return false;
  }
  const repository = repositoryContainer.getRevokedTokenRepository();
  await repository.createOne({ token, revokedAt: new Date() });
  debug('Signout token: %s', token);
  return true;
};

// サインアウト済みか検証
export const isSignedout = async (token?: string | null): Promise<boolean> => {
  if (!token) {
    return false;
  }
  const repository = repositoryContainer.getRevokedTokenRepository();
  return !!(await repository.findOne({ token }));
};

File: packages/golang/domains/auth/signout.go
package auth

import (
	"context"
	"time"

	"github.com/cam-inc/viron/packages/golang/repositories"

	"github.com/cam-inc/viron/packages/golang/repositories/container"
)

func SignOut(ctx context.Context, token string) bool {
	if token == "" {
		return false
	}

	repo := container.GetRevokedTokensRepository()
	revokedToken := &repositories.RevokedTokenEntity{
		Token:     token,
		RevokedAt: time.Now(),
	}

	if _, err := repo.CreateOne(ctx, revokedToken); err != nil {
		return false
	}

	return true

}

func IsSignedOut(ctx context.Context, token string) bool {
	// tokenが空の場合はsingoutされていないと判断
	if token == "" {
		return false
	}
	repo := container.GetRevokedTokensRepository()
	entity, err := repo.FindOne(ctx, token)
	if err != nil {
		return false
	}
	revoked := &repositories.RevokedTokenEntity{}
	if err := entity.Bind(revoked); err != nil {
		return false
	}
	return true
}
